### PR TITLE
Windows start scripts use pathing jar instead of explicit classpath

### PR DESCRIPTION
**Pull request by uschi2000**
[Review on Reviewable](https://reviewable.io/reviews/palantir/gradle-java-distribution/25)

### CODE REVIEW

**Comment by uschi2000** [comment]
Take a look please, @markelliot . This resolves a common problem where in generated Windows start scripts the CLASSPATH variable is too long (too long for Windows, that is).

**Comment by markelliot** [comment]
(summarizing our offline chat) I'd prefer to find a way to do this in the batch script rather than embedding the classpath in the primary jar, as that way new jars may be added in-order without having to unpack/repack jars. If that can't be done, we can go with this as a workaround.

**Comment by uschi2000** [comment]
The classpath is already a variable in the generated bat script:

```
set CLASSPATH=%APP_HOME%\lib\foundry-cli-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\multipass-extensions-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\remote-clients-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\foundry-catalog-api-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\foundry-fsclient-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\foundry-fsproxy-api-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\foundry-metadata-api-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\logback-classic-1.1.3.jar;%APP_HOME%\lib\jackson-dataformat-yaml-2.6.1.jar;%APP_HOME%\lib\jackson-datatype-guava-2.6.1.jar;%APP_HOME%\lib\jackson-datatype-jdk7-2.6.1.jar;%APP_HOME%\lib\retrofit-1.9.0.jar;%APP_HOME%\lib\joda-time-2.8.2.jar;%APP_HOME%\lib\commons-lang3-3.3.2.jar;%APP_HOME%\lib\jcl-over-slf4j-1.7.12.jar;%APP_HOME%\lib\slf4j-api-1.7.12.jar;%APP_HOME%\lib\spring-shell-1.1.1-palantir.jar;%APP_HOME%\lib\guava-18.0.jar;%APP_HOME%\lib\jackson-databind-2.6.1.jar;%APP_HOME%\lib\jsr305-3.0.0.jar;%APP_HOME%\lib\feign-jackson-8.10.0.jar;%APP_HOME%\lib\feign-jaxrs-8.10.0.jar;%APP_HOME%\lib\feign-okhttp-8.10.0.jar;%APP_HOME%\lib\javax.ws.rs-api-2.0.1.jar;%APP_HOME%\lib\foundry-api-shared-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\resource-identifier-0.3.0.jar;%APP_HOME%\lib\foundry-build-api-develop-454-g7e9ab3a46679-dirty.jar;%APP_HOME%\lib\hadoop-s3-adapter-0.5.18-dev.1+g814be3c.b153.jar;%APP_HOME%\lib\hadoop-client-2.7.1.jar;%APP_HOME%\lib\hadoop-aws-2.7.1.jar;%APP_HOME%\lib\jersey-common-2.19.jar;%APP_HOME%\lib\javax.servlet-api-3.1.0.jar;%APP_HOME%\lib\logback-core-1.1.3.jar;%APP_HOME%\lib\jackson-core-2.6.1.jar;%APP_HOME%\lib\snakeyaml-1.15.jar;%APP_HOME%\lib\gson-2.2.4.jar;%APP_HOME%\lib\spring-core-4.0.6.RELEASE.jar;%APP_HOME%\lib\spring-context-support-4.0.6.RELEASE.jar;%APP_HOME%\lib\commons-io-2.4.jar;%APP_HOME%\lib\jline-2.11.jar;%APP_HOME%\lib\jackson-annotations-2.6.1.jar;%APP_HOME%\lib\feign-core-8.10.0.jar;%APP_HOME%\lib\okhttp-2.5.0.jar;%APP_HOME%\lib\hadoop-common-2.7.1.jar;%APP_HOME%\lib\hadoop-hdfs-2.7.1.jar;%APP_HOME%\lib\hadoop-mapreduce-client-app-2.7.1.jar;%APP_HOME%\lib\hadoop-yarn-api-2.7.1.jar;%APP_HOME%\lib\hadoop-mapreduce-client-core-2.7.1.jar;%APP_HOME%\lib\hadoop-mapreduce-client-jobclient-2.7.1.jar;%APP_HOME%\lib\hadoop-annotations-2.7.1.jar;%APP_HOME%\lib\aws-java-sdk-1.7.4.jar;%APP_HOME%\lib\javax.annotation-api-1.2.jar;%APP_HOME%\lib\jersey-guava-2.19.jar;%APP_HOME%\lib\hk2-api-2.4.0-b25.jar;%APP_HOME%\lib\javax.inject-2.4.0-b25.jar;%APP_HOME%\lib\hk2-locator-2.4.0-b25.jar;%APP_HOME%\lib\osgi-resource-locator-1.0.1.jar;%APP_HOME%\lib\spring-beans-4.0.6.RELEASE.jar;%APP_HOME%\lib\spring-context-4.0.6.RELEASE.jar;%APP_HOME%\lib\okio-1.6.0.jar;%APP_HOME%\lib\commons-cli-1.2.jar;%APP_HOME%\lib\commons-math3-3.4.1.jar;%APP_HOME%\lib\xmlenc-0.52.jar;%APP_HOME%\lib\commons-httpclient-3.1.jar;%APP_HOME%\lib\commons-codec-1.6.jar;%APP_HOME%\lib\commons-net-3.1.jar;%APP_HOME%\lib\commons-collections-3.2.1.jar;%APP_HOME%\lib\jetty-6.1.26.jar;%APP_HOME%\lib\jetty-util-6.1.26.jar;%APP_HOME%\lib\commons-logging-1.1.3.jar;%APP_HOME%\lib\log4j-1.2.17.jar;%APP_HOME%\lib\jets3t-0.9.0.jar;%APP_HOME%\lib\commons-lang-2.6.jar;%APP_HOME%\lib\commons-configuration-1.6.jar;%APP_HOME%\lib\jackson-core-asl-1.9.13.jar;%APP_HOME%\lib\jackson-mapper-asl-1.9.13.jar;%APP_HOME%\lib\avro-1.7.7.jar;%APP_HOME%\lib\protobuf-java-2.6.1.jar;%APP_HOME%\lib\hadoop-auth-2.7.1.jar;%APP_HOME%\lib\jsch-0.1.42.jar;%APP_HOME%\lib\curator-client-2.7.1.jar;%APP_HOME%\lib\curator-recipes-2.7.1.jar;%APP_HOME%\lib\htrace-core-3.1.0-incubating.jar;%APP_HOME%\lib\zookeeper-3.4.6.jar;%APP_HOME%\lib\commons-compress-1.4.1.jar;%APP_HOME%\lib\jsp-api-2.1.jar;%APP_HOME%\lib\netty-all-4.0.29.Final.jar;%APP_HOME%\lib\xercesImpl-2.9.1.jar;%APP_HOME%\lib\leveldbjni-all-1.8.jar;%APP_HOME%\lib\hadoop-mapreduce-client-common-2.7.1.jar;%APP_HOME%\lib\hadoop-mapreduce-client-shuffle-2.7.1.jar;%APP_HOME%\lib\hadoop-yarn-common-2.7.1.jar;%APP_HOME%\lib\httpclient-4.2.5.jar;%APP_HOME%\lib\hk2-utils-2.4.0-b25.jar;%APP_HOME%\lib\aopalliance-repackaged-2.4.0-b25.jar;%APP_HOME%\lib\javassist-3.18.1-GA.jar;%APP_HOME%\lib\spring-aop-4.0.6.RELEASE.jar;%APP_HOME%\lib\spring-expression-4.0.6.RELEASE.jar;%APP_HOME%\lib\httpcore-4.2.4.jar;%APP_HOME%\lib\java-xmlbuilder-0.4.jar;%APP_HOME%\lib\commons-digester-1.8.jar;%APP_HOME%\lib\commons-beanutils-core-1.8.0.jar;%APP_HOME%\lib\paranamer-2.3.jar;%APP_HOME%\lib\snappy-java-1.0.5.jar;%APP_HOME%\lib\apacheds-kerberos-codec-2.0.0-M15.jar;%APP_HOME%\lib\curator-framework-2.7.1.jar;%APP_HOME%\lib\netty-3.8.0.Final.jar;%APP_HOME%\lib\xz-1.0.jar;%APP_HOME%\lib\xml-apis-1.3.04.jar;%APP_HOME%\lib\hadoop-yarn-client-2.7.1.jar;%APP_HOME%\lib\hadoop-yarn-server-common-2.7.1.jar;%APP_HOME%\lib\hadoop-yarn-server-nodemanager-2.7.1.jar;%APP_HOME%\lib\jaxb-api-2.2.2.jar;%APP_HOME%\lib\jersey-client-1.9.jar;%APP_HOME%\lib\jackson-jaxrs-1.9.13.jar;%APP_HOME%\lib\jackson-xc-1.9.13.jar;%APP_HOME%\lib\guice-3.0.jar;%APP_HOME%\lib\jersey-guice-1.9.jar;%APP_HOME%\lib\javax.inject-1.jar;%APP_HOME%\lib\aopalliance-1.0.jar;%APP_HOME%\lib\commons-beanutils-1.7.0.jar;%APP_HOME%\lib\apacheds-i18n-2.0.0-M15.jar;%APP_HOME%\lib\api-asn1-api-1.0.0-M20.jar;%APP_HOME%\lib\api-util-1.0.0-M20.jar;%APP_HOME%\lib\jettison-1.1.jar;%APP_HOME%\lib\stax-api-1.0-2.jar;%APP_HOME%\lib\activation-1.1.jar;%APP_HOME%\lib\cglib-2.2.1-v20090111.jar;%APP_HOME%\lib\asm-3.3.1.jar

@rem Execute foundry-cli
"%JAVA_EXE%" %DEFAULT_JVM_OPTS% %JAVA_OPTS% %FOUNDRY_CLI_OPTS%  -classpath "%CLASSPATH%" com.palantir.foundry.cli.main.FoundryCli %CMD_LINE_ARGS%
```

The only other way I know is to use lib/\* as classpath, but it means the jar order is not stable.

**Comment by uschi2000** [comment]
Note that something's not quite working right, so if we want the method of this PR, I'll have to fix this up before merging.

**Comment by markelliot** [comment]
Ugh, lame. We should file a but on Gradle for this too, since we're really just wrapping the standard Gradle tasks and I'd really like to just keep depending on those. 

Still considering what options we have here...but looks like this may be only workaround.

**Comment by uschi2000** [comment]
There's another work around, mention in a Gradle forum: https://discuss.gradle.org/t/gradle-application-plugin-and-cmd-exe-line-size-limitations/2521

**Comment by uschi2000** [comment]
Also, this is a bug already: https://issues.gradle.org/browse/GRADLE-2992

**Comment by markelliot** [comment]
Ok, so I like the implicit CLASSPATH solution that gets us to 32k chars. Is 32k chars enough for the thing we're having a problem with?

If so, we can specify the way the batch file is constructed easily enough as part of the build.

**Comment by uschi2000** [comment]
I tried that solution and it does not seem to work, at least not on Windows 7. The limit is somewhere around 8100 bytes in my test.

**Comment by uschi2000** [comment]
(and our current classpath is about 10k if executed from C:\Users/rfink/Desktop\a\

**Comment by markelliot** [comment]
https://docs.oracle.com/javase/tutorial/essential/environment/paths.html says that setting the env var _should_ work, not sure why it isn't. 

http://superuser.com/questions/387619/overcoming-the-1024-character-limit-with-setx, which links to https://msdn.microsoft.com/en-us/library/windows/desktop/ms686206(v=vs.85).aspx ...suggests 32,767 should be the single variable limit. 

Sent from my iPhone

File: src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy [hunk 1]
```diff
@@ -29,9 +31,28 @@ class JavaDistributionPlugin implements Plugin<Project> {
...
+                def winScriptFile = project.file getWindowsScript()
+                def winFileText = winScriptFile.text
+
+                // Remove too-long-classpath and use pathing jar instead
+                winFileText = winFileText.replaceAll('set CLASSPATH=.*', 'rem CLASSPATH declaration removed.')
+                winFileText = winFileText.replaceAll(
+                    '("%JAVA_EXE%" .* -classpath ")%CLASSPATH%(" .*)',
+                    '$1%APP_HOME%\\\\lib\\\\' + pathingJar.archiveName + '$2')
+
+                winScriptFile.text = winFileText
```
Comment (markelliot) [inline src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy:55]: i'm tempted for us to bundle our own template batch file here as we can, as part of the CreateStartScripts task, override the default one Gradle uses.

File: src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy [hunk 2]
```diff
@@ -29,9 +31,28 @@ class JavaDistributionPlugin implements Plugin<Project> {
 
         DistributionExtension ext = project.extensions.create('distribution', DistributionExtension)
 
+        // Specify classpath using pathing jar rather than command line argument on Windows, since
+        // Windows path sizes are limited.
+        PathingJarTask pathingJar = project.tasks.create("pathingJar", PathingJarTask)
```
Comment (markelliot) [inline src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy]: this should have an `onlyIf ext.enablePathingJar`, most likely

**Comment by markelliot** [comment]
is this thing commonly called a 'pathing' jar? would classpathing or manifest be better terms?

File: src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy (hunk 1 above)
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy:55]: leaving as it stands for now, as per offline discussion.

File: src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy (hunk 2 above)
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/javadist/JavaDistributionPlugin.groovy]: +closure, done.

**Comment by uschi2000** [comment]
renamed to "manifest classpath jar"

**Comment by uschi2000** [comment]
Take another look please, @markelliot

File: src/main/groovy/com/palantir/gradle/javadist/ManifestClasspathJarTask.groovy [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class ManifestClasspathJarTask extends Jar {
+
+    public ManifestClasspathJarTask() {
+        appendix = 'manifest-classpath'
+    }
+
+    public void configure(DistributionExtension ext) {
+        manifest.attributes("Class-Path": project.files(project.configurations.runtime)
+            .collect { println(it.getName()); return it.getName() }
```
Comment (markelliot) [inline src/main/groovy/com/palantir/gradle/javadist/ManifestClasspathJarTask.groovy]: is this println supposed to be here?

File: src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy [hunk 4]
```diff
@@ -210,6 +210,26 @@ class JavaDistributionPluginTests extends Specification {
...
+        buildResult.task(':untar').outcome == TaskOutcome.SUCCESS
+
+        new File(projectDir, 'dist/service-name-0.1/service/bin/service-name.bat').exists()
+        String startScript = readFully('dist/service-name-0.1/service/bin/service-name.bat')
+        startScript.contains("-manifest-classpath-0.1.jar")
+        !startScript.contains("-classpath \"%CLASSPATH%\"")
+        new File(projectDir, 'dist/service-name-0.1/service/lib/').listFiles()
+            .find({it.name.endsWith("-manifest-classpath-0.1.jar")})
+    }
+
```
Comment (markelliot) [inline src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy:237]: suspect we should add a test where `enableManifestClasspath` is false that the jar is not produced and that the manifestClasspathJar task is skipped

Comment (uschi2000) [inline src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy:237]: done

**Pull request merged**